ios/browser/api/sync/brave_sync_api.mm

```diff
@@ -0,0 +1,203 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#import "brave/ios/browser/api/sync/brave_sync_api.h"
+#import "brave/ios/browser/api/sync/brave_sync_worker.h"
+
+#import <CoreImage/CoreImage.h>
+#include <string>
+#include <vector>
+
+#include "base/json/json_writer.h"
+#include "base/memory/weak_ptr.h"
+#include "base/scoped_observer.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/sys_string_conversions.h"
+#include "brave/components/brave_sync/brave_sync_prefs.h"
+#include "brave/components/brave_sync/crypto/crypto.h"
+#include "components/sync/driver/profile_sync_service.h"
+#include "components/sync/driver/sync_service.h"
+#include "components/sync/driver/sync_service_observer.h"
+#include "components/sync_device_info/device_info.h"
+#include "components/sync_device_info/device_info_sync_service.h"
+#include "components/sync_device_info/device_info_tracker.h"
+#include "components/sync_device_info/local_device_info_provider.h"
+#include "ios/chrome/browser/application_context.h"
+#include "ios/chrome/browser/browser_state/chrome_browser_state.h"
+#include "ios/chrome/browser/browser_state/chrome_browser_state_manager.h"
+#include "ios/chrome/browser/sync/device_info_sync_service_factory.h"
+#include "ios/chrome/browser/sync/profile_sync_service_factory.h"
+#include "ios/chrome/browser/sync/sync_setup_service.h"
+#include "ios/chrome/browser/sync/sync_setup_service_factory.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+
+
+@interface BraveSyncDeviceObserver()
+{
+  std::unique_ptr<BraveSyncDeviceTracker> _device_observer;
+}
+@end
+
+@implementation BraveSyncDeviceObserver
+
+- (instancetype)initWithCallback:(void(^)())onDeviceInfoChanged {
+  if ((self = [super init])) {
+    _device_observer = std::make_unique<BraveSyncDeviceTracker>(onDeviceInfoChanged);
+  }
+  return self;
+}
+@end
+
+@interface BraveSyncServiceObserver()
+{
+  std::unique_ptr<BraveSyncServiceTracker> _service_observer;
+}
+@end
+
+@implementation BraveSyncServiceObserver
+
+- (instancetype)initWithCallback:(void(^)())onSyncServiceStateChanged {
+  if ((self = [super init])) {
+    _service_observer = std::make_unique<BraveSyncServiceTracker>([onSyncServiceStateChanged](syncer::SyncService *sync) {
+      onSyncServiceStateChanged();
+    }, [](syncer::SyncService *sync) {
+        fprintf(stderr, "Sync Shut Down\n");
+    });
+  }
+  return self;
+}
+@end
+
+@interface BraveSyncAPI()
+{
+  std::unique_ptr<BraveSyncWorker> _worker;
+}
+@end
+
+@implementation BraveSyncAPI
+
++ (instancetype)sharedSyncAPI {
+	static BraveSyncAPI *instance = nil;
+	static dispatch_once_t onceToken;
+	dispatch_once(&onceToken, ^{
+		instance = [[BraveSyncAPI alloc] init];
+	});
+	return instance;
+}
+
+- (instancetype)init {
+  if ((self = [super init])) {
+    ios::ChromeBrowserStateManager* browserStateManager =
```

Comment:
this should move into sharedSyncAPI and and pass the browser state into the constructor